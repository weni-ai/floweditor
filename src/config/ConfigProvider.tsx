import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR =
  'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<
  ConfigProviderProps
> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType,
  };

  constructor(props: ConfigProviderProps) {
    super(props);

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config,
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
