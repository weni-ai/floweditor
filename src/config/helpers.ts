import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  Types,
  VISIBILITY_ONLINE,
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';
import i18n from './i18n';
import { NodeEditorSettings } from '../store/nodeEditor';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (
  operators: Operator[],
  config: FlowEditorConfig,
): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (
  typeConfigs: Type[],
  config: FlowEditorConfig,
  nodeSettings: NodeEditorSettings,
): Type[] => {
  return filterVisibility(typeConfigs, config, nodeSettings).map(
    (typeConfig: Type) => {
      if (typeConfig.new) {
        return {
          ...typeConfig,
          description: `${typeConfig.description} (${i18n.t('new', 'New')})`,
        };
      }

      return typeConfig;
    },
  );
};

const filterVisibility = (
  items: FlowTypeVisibility[],
  config: FlowEditorConfig,
  nodeSettings?: NodeEditorSettings,
): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    if (nodeSettings) {
      const typeItem = item as Type;
      if (typeItem.type === Types.call_brain) {
        // check if we can show the call brain action
        // if we are not a ghost node
        // and if we are in a multi action node, we can't show the call brain action
        if (
          nodeSettings.originalNode.node.actions.length >= 1 &&
          !nodeSettings.originalNode.ghost
        ) {
          console.log('filter', nodeSettings);
          return false;
        }
      }
    }

    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (
        !(config.filters || []).find((name: string) => name === item.filter)
      ) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return (
      item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1
    );
  });
};
