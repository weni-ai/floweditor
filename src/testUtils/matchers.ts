import { commaListsOr } from 'common-tags';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

const snapshot = require('jest-snapshot');

const matchers: jest.ExpectExtendMap = {};

interface MatchResult {
  message: () => string;
  pass: boolean;
}

function toBeUnique<T>(this: jest.MatcherUtils, received: any[]): MatchResult {
  const seen: { [value: string]: boolean } = {};

  for (const item of received) {
    const stringified = JSON.stringify(item, null, 1);
    if (seen[stringified]) {
      return {
        message: () => `Duplicate item in array:\n${stringified}`,
        pass: false,
      };
    } else {
      seen[stringified] = true;
    }
  }

  return {
    message: () => 'Array is unique',
    pass: true,
  };
}

function toMatchCallSnapshot<T>(
  this: jest.MatcherUtils,
  received: any,
  snapshotName?: string,
): MatchResult {
  return snapshot.toMatchSnapshot.call(
    this,
    received.mock.calls[0],
    snapshotName,
  );
}

function toHaveInboundConnections<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
): MatchResult {
  if (Object.keys(received.inboundConnections).length > 1) {
    return {
      message: () => `Node ${received.node.uuid} has an inbound connection`,
      pass: true,
    };
  }

  return {
    message: () => `Node ${received.node.uuid} has no inbound connections`,
    pass: false,
  };
}

function toHaveInboundFrom<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
  expected: Exit,
): MatchResult {
  for (const exitUUID of Object.keys(received.inboundConnections)) {
    if (exitUUID === expected.uuid) {
      return {
        message: () =>
          `Node ${received.node.uuid} incorrectly has inbound from ${expected.uuid}`,
        pass: true,
      };
    }
  }

  return {
    message: () =>
      `Node ${received.node.uuid} doesn't have inbound connection from ${expected.uuid}`,
    pass: false,
  };
}

function toPointTo<T>(
  this: jest.MatcherUtils,
  received: Exit,
  expected: RenderNode,
): MatchResult {
  if (received.destination_uuid === expected.node.uuid) {
    return {
      message: () =>
        `${received.uuid} incorrectly points to ${expected.node.uuid}`,
      pass: true,
    };
  }

  return {
    message: () =>
      `Expected ${received.uuid} to point to ${expected.node.uuid}`,
    pass: false,
  };
}

function toHaveExitWithDestination<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
): MatchResult {
  for (const exit of received.node.exits) {
    if (exit.destination_uuid !== null) {
      return {
        message: () =>
          `Exit ${exit.uuid} in node ${received.node.uuid} has a destination`,
        pass: true,
      };
    }
  }

  return {
    message: () => `All exits for ${received.node.uuid} are disconnected`,
    pass: false,
  };
}

function toHaveExitThatPointsTo<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
  expected: RenderNode,
): MatchResult {
  for (const exit of received.node.exits) {
    if (exit.destination_uuid === expected.node.uuid) {
      return {
        message: () =>
          `${exit.uuid} incorrectly has exit that points to ${expected.node.uuid}`,
        pass: true,
      };
    }
  }

  return {
    message: () =>
      `Expected ${received.node.uuid} to have an exit that points to ${expected.node.uuid}`,
    pass: false,
  };
}

function toHavePayload<T>(
  this: jest.MatcherUtils,
  store: any,
  actionType: string,
  expectedPayload: any,
): MatchResult {
  const payload = JSON.stringify(expectedPayload);
  for (const actionTaken of store.getActions()) {
    if (actionTaken.type === actionType) {
      if (JSON.stringify(actionTaken.payload) === payload) {
        return {
          message: () =>
            `Result contained action type ${actionType} with payload ${payload}`,
          pass: true,
        };
      }
    }
  }

  return {
    message: () =>
      `Could not find action type ${actionType} with payload ${payload}`,
    pass: false,
  };
}

function toHaveReduxActions<T>(
  this: jest.MatcherUtils,
  store: any,
  actionTypes: string[],
): MatchResult {
  const missedTypes: any[] = [];
  for (const actionTaken of store.getActions()) {
    for (const actionType of actionTypes) {
      if (actionTaken.type === actionType) {
        return {
          message: () => `Result contained action type ${actionType}`,
          pass: true,
        };
      } else {
        missedTypes.push(actionType);
      }
    }
  }

  return {
    message: () =>
      commaListsOr`Could not find action type(s): ${[...missedTypes]}`,
    pass: false,
  };
}

expect.extend({
  toPointTo,
  toHaveExitThatPointsTo,
  toHaveInboundFrom,
  toHaveExitWithDestination,
  toHaveInboundConnections,
  toHavePayload,
  toHaveReduxActions,
  toMatchCallSnapshot,
  toBeUnique,
});
