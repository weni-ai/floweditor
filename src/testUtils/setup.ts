/* eslint-disable no-redeclare */
import { Console } from 'console';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

// Declare custom matcher types
declare global {
  namespace jest {
    interface Matchers<R> {
      toPointTo(renderNode: RenderNode): R;
      toHaveExitThatPointsTo(renderNode: RenderNode): R;
      toHaveInboundFrom(exit: Exit): R;
      toHaveExitWithDestination(): R;
      toHaveInboundConnections(): R;
      toHavePayload(action: string, payload: any): R;
      toHaveReduxActions(actions: string[]): R;
      toMatchCallSnapshot(snapshotName?: string): R;
    }
  }
}

// Ensure console logs are visible while running tests https://github.com/facebook/jest/issues/3853
global.console = new Console(process.stderr, process.stderr);

// Configure Enzyme adapter
configure({ adapter: new Adapter() });

// RAF shim
// tslint:disable-next-line:ban-types
(global as any).requestAnimationFrame = (callback: Function) => {
  setTimeout(callback, 0);
};
