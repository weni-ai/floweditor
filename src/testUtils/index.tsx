// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
