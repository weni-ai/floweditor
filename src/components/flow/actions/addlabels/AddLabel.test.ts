import AddLabelsComp, {
  MAX_TO_SHOW,
} from 'components/flow/actions/addlabels/AddLabels';
import { Types } from 'config/interfaces';
import { AddLabels } from 'flowTypes';
import { composeComponentTestUtils } from 'testUtils';

const labels = [
  'Help',
  'New',
  'Feedback',
  'Needs Attention',
  'Running Out of Plausible Label Names',
  'But alas, here is another one',
];

const baseProps: AddLabels = {
  type: Types.add_input_labels,
  uuid: `${Types.add_input_labels}-0`,
  labels: labels.map((name, idx) => ({ name, uuid: `label-${idx}` })),
};

const { setup } = composeComponentTestUtils(AddLabelsComp, baseProps);

describe(AddLabelsComp.name, () => {
  it('should display labels on action', () => {
    const { wrapper } = setup();

    // Assert that we're displaying the max labels
    // we want to display plus an ellipses.
    // the plus one is to include de center wrapper
    expect(wrapper.find('div').length).toBe(MAX_TO_SHOW + 1);
    expect(wrapper).toMatchSnapshot();
  });
});
