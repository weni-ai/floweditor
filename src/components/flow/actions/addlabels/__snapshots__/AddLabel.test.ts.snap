// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddLabelsComp > should display labels on action 1`] = `
<Fragment>
  <div
    style={
      {
        "textAlign": "center",
      }
    }
  >
    <div
      className="_node_asset_9658e1"
      key="label-0"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      Help
    </div>
    <div
      className="_node_asset_9658e1"
      key="label-1"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      New
    </div>
    <div
      className="_node_asset_9658e1"
      key="label-2"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      Feedback
    </div>
    <div
      className="_node_asset_9658e1"
      key="label-3"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      Needs Attention
    </div>
    <div
      key="ellipses"
    >
      +
      2
       more
    </div>
  </div>
</Fragment>
`;

exports[`AddLabelsComp > should display labels that contains name_match key 1`] = `
<Fragment>
  <div
    style={
      {
        "textAlign": "center",
      }
    }
  >
    <div
      className="_node_asset_9658e1"
      key="Look at me, i have a name_match key"
    >
      Look at me, i have a name_match key
    </div>
    <div
      className="_node_asset_9658e1"
      key="label-0"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      Help
    </div>
    <div
      className="_node_asset_9658e1"
      key="label-1"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      New
    </div>
    <div
      className="_node_asset_9658e1"
      key="label-2"
    >
      <ForwardRef
        className="_icon_25b42d"
        icon="bookmark-tag-1"
        scheme="neutral-dark"
        size="avatar-nano"
      />
      Feedback
    </div>
    <div
      key="ellipses"
    >
      +
      3
       more
    </div>
  </div>
</Fragment>
`;
